### ISSUE TITLE
Implement Prefix-LM attention masking

### ISSUE THREAD

**Issue by TokyoExpress** [EleutherAI/gpt-neox#805]
relevant https://github.com/bigscience-workshop/Megatron-DeepSpeed/issues/21

I'm trying to implement the attention mask for Prefix-LM as in the T5 paper in order to do instruction finetuning.
![image](https://user-images.githubusercontent.com/43324265/222028899-b863196e-be9f-45b5-aeb5-5633476717b3.png)

I am having trouble modifying the attention masking operation as it seems to be part of the fused kernels and I cannot identify which part of the C++ code is responsible for applying the mask.

One main issue is that a different attention mask will be needed for every item in the batch (depending on the length of the prompt), while the current code only supports a single attention mask for the entire batch.

Could someone point me in the right direction here?

**Comment by TokyoExpress** [comment]
Essentially I would like to unzero-out specific chunks in each sequence so that the prompt can undergo full attention visibility before beginning to decode the output during training. This would be done by taking the spans in the token sequence that correspond to prompts (there may be multiple due to sequence packing, so essentially `<|endoftext|>` to the end of the prompt) and applying squares of 1's to the lower triangular matrix to achieve the desired attention mask.

I have no trouble creating these masks given my inputs, I'm just not sure how to get these masks into the forward operation.
